accepted/2021/platform-dependencies/platform-dependencies.md

```diff
@@ -0,0 +1,433 @@
+
+* There is no dependency data that is common to all patch releases. While there could end up being commonality in the data, it is not an inherent feature.
+* From a human-readability standpoint, the size of the json file would be quite large and difficult to navigate as patch releases accumulate.
+
+## Engineering Workflow
+
+It is necessary for all .NET contributors to participate in the maintenance of the platform dependency model. The scope of .NET is too large to have a single person or small set of people actively monitor for changes to the dependencies. It needs to be a distributed effort where the maintainers that are familiar with that area of .NET can keep an eye on things.
+
+There will also be a small group charged with maintaining the platform dependency tracking system as a whole. This group will be responsible for approving updates to the model and acting as points-of-contact for any questions from contributors.
+
+Luckily, the burden of maintenance is expected to be low since dependencies don't often change. However, there are a few scenarios that do cause change which are described further below.
+
+### Maintenance Scenarios
+
+In all of the scenarios below, the person introducing or discovering the dependency change should file an issue in the [dotnet/core](https://github.com/dotnet/core) repo describing the change. Such issues should be labeled as `area-dependency`. This is necessary in order to notify stakeholder teams (e.g. installers, container images, documentation) of such changes.
+
+#### New Dependency
+
+New platform dependencies can be introduced by making a code change that has the dependency or including a new component into .NET (e.g. shared framework, NuGet package). When such a dependency is added, the contributor should also update the appropriate platform dependency model file with the necessary information that describes the dependency across all supported platforms.
+
+##### Change Detection
+
+In order to avoid the reliance upon contributors to recognize when they've changed a dependency, a more automated solution would be preferable. This can be done by defining a GitHub bot that checks for files in PRs containing `NativeMethods` or `Interop` in their name. If such a file is detected, a label is added to the PR alerting the submitter that they should evaluate their changes for changes to the dependencies. While not a fool-proof solution, it should provide coverage for the vast majority of dependencies. Work is still required by the submitter to make the appropriate changes to the platform dependency model but the GitHub bot helps alert them when there is potentially action that is needed.
```

Comment:
I don't think that this is sufficient.  To help keep dependency tracking live we will need to invest more deeply in dependency identification.  Also how would we ensure that toolchain dependencies are being identified automatically?

Reply:
I could imagine something like - run all the tests under `ltrace` or similar, and compare all the dlopen's seen to a baseline list. When a new one is seen, it's an indication that someone should check whether a json update is needed, then update the baseline.

It seems to me that isn't part of 'crawl' but maybe 'run'

Reply:
This is probably the aspect of this proposal that will require the most iteration and refinement over time.  I don't want to over-invest in this, however. I fully admit that the proposed change detection isn't foolproof.  But it's a fairly cheap and non-disruptive solution to get started with.  I think as we learn of other areas that require detection, we can work to address those as needed.

I do like the suggestion of @danmoseley for an advanced implementation of detection logic.  This could apply to both runtime and toolchain dependencies.

Reply:
It needs to be complete to the extent that we're aware of the dependencies.  The model is intended to be used as input to other dependent artifacts like Dockerfiles, as you suggest (see phases 2 & 4 in https://github.com/dotnet/core/issues/5651 for others).  I don't see the ability to know and describe the dependencies as being unfeasible.  Indeed, it better not be; otherwise, what are we documenting for customers?  The change detection aspect is certainly a challenge and I feel should be scoped, at least to begin with, to be able to detect most dependency changes rather than all.

My view is that we treat the dependency model as a work in progress for the lifetime of the .NET version it's associated with.  We do our best effort to define the model and if we discover something later on several releases later, we can just go back and edit it to accurately reflect what is known.  The benefit of having downstream assets consuming this model data is that it acts a form of validation.  If the Dockerfiles are synced with the dependency model and something doesn't work in the container because it's missing a dependency, then we know the model isn't correct.  I think we just keep iterating on it until it's right.

Reply:
Ah, somehow I had thought you were trying to describe the closure, ie., including transitive dependencies. Yes, just describing direct dependencies is surely doable.